File: ABC/ABC336/F/main.py
import queue

def search(s, swap_lists):
    ret = dict()
    ret[s] = 0
    que = queue.Queue()
    que.put(s)
    while que:
        v = que.get()
        dist = ret[v]
        if dist == 10: break
        for swap_list in swap_lists:
            nv = list(v)
            for x, y in swap_list:
                nv[x] = v[y]
            nv = tuple(nv)
            if not nv in ret:
                ret[nv] = dist + 1
                que.put(nv)

    return ret

h, w = map(int, input().split())
n = h * w

s = [0] * n
t = [i for i in range(n)]

for i in range(h):
    tmp = list(map(int, input().split()))
    for j in range(w):
        s[i * w + j] = tmp[j] - 1
s = tuple(s)
t = tuple(t)

swap_lists = []
for sr in range(2):
    for sc in range(2):
        swap_list = []
        for r in range(sr, h - 1 + sr):
            for c in range(sc, w - 1 + sc):
                swap_list.append((r * w + c, (sr + sr + h - 2 - r) * w + sc + sc + w - 2 - c))
        swap_lists.append(swap_list)



dic1 = search(s, swap_lists)
dic2 = search(t, swap_lists)

res = 30
for k in dic1:
    if k in dic2:
        res = min(res, dic1[k] + dic2[k])

print(res if res < 30 else -1)

File: ABC/ABC336/F/main.cc
#include <iostream>
#include <map>
#include <numeric>
#include <queue>
#include <vector>

using namespace std;

class Grid {
 public:
  Grid(const int &height, const int &width) : height_(height), width_(width) { }
  int To1d(const int &row, const int &column) const { return row * width_ + column; }
  pair<int, int> To2d(const int &v) const { return make_pair(v / width_, v % width_); }

 private:
  int height_, width_;
};

map<vector<int>, int> Search(const vector<int> &s, const vector<vector<pair<int, int> > > &swap_lists) {
  map<vector<int>, int> mp;
  mp[s] = 0;
  queue<vector<int> > que;
  que.emplace(s);
  while (!que.empty()) {
    auto v = que.front();
    que.pop();
    int dist = mp[v];
    if (dist == 10) break;
    for (const auto &swap_list : swap_lists) {
      auto nv = v;
      for (auto [x, y] : swap_list) {
        nv[x] = v[y];
      }
      if (!mp.count(nv)) {
        mp[nv] = dist + 1;
        que.emplace(nv);
      }
    }
  }
  return mp;
}

int main() {
  int h, w;
  cin >> h >> w;
  const int n = h * w;
  Grid grid(h, w);
  vector<int> s(n), t(n);
  for (auto &e : s) { cin >> e; --e; }
  iota(t.begin(), t.end(), 0);

  vector<vector<pair<int, int> > > swap_lists;
  for (int sr = 0; sr < 2; ++sr) {
    for (int sc = 0; sc < 2; ++sc) {
      vector<pair<int, int> > swap_list;
      for (int r = sr; r < h - 1 + sr; ++r) {
        for (int c = sc; c < w - 1 + sc; ++c) {
          swap_list.emplace_back(grid.To1d(r, c), grid.To1d(sr + sr + h - 2 - r, sc + sc + w - 2 - c));
        }
      }
      swap_lists.emplace_back(swap_list);
    }
  }

  auto mp1 = Search(s, swap_lists);
  auto mp2 = Search(t, swap_lists);

  int res = 30;
  for (auto m : mp1) {
    if (mp2.count(m.first)) {
      res = min(res, m.second + mp2[m.first]);
    }
  }
  cout << (res < 30 ? res : -1) << endl;

  return 0;
}